client/state/themes/upload-theme/reducer.js

```diff
@@ -0,0 +1,71 @@
+/**
+ * External dependencies
+ */
+import { combineReducers } from 'redux';
+import { omit } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { createReducer } from 'state/utils';
+import {
+	THEME_UPLOAD_START,
+	THEME_UPLOAD_SUCCESS,
+	THEME_UPLOAD_FAILURE,
+	THEME_UPLOAD_CLEAR,
+	THEME_UPLOAD_PROGRESS,
+} from 'state/action-types';
+
+const clearState = function( state, { siteId } ) {
+	return omit( state, siteId );
+};
+
+// Returns a func that sets state to be the specified action property,
+// or true if no property supplied
+const setState = function( actionProp ) {
+	return ( state, action ) => {
+		const { siteId } = action;
+		const newState = Object.assign( {}, state );
+		newState[ siteId ] = actionProp ? action[ actionProp ] : true;
+		return newState;
+	};
+};
+
+export const uploadedTheme = createReducer( {}, {
+	[ THEME_UPLOAD_SUCCESS ]: setState( 'theme' ),
+	[ THEME_UPLOAD_CLEAR ]: clearState,
+	[ THEME_UPLOAD_START ]: clearState,
```

Comment:
The recurring `THEME_UPLOAD_CLEAR` and `THEME_UPLOAD_START` handlers are a bit puzzling to me. I've had another look at #9632, and I wonder -- do we even need the `clearThemeUpload` action at all? They seem be mostly about resetting upload state on a new page load. But if e.g. an upload (of a very big file) is still ongoing for a site, won't the progress callback keep updating the state anyway? (And from a UX POV, it might be okay to do so -- it's probably an edge case anyway.)

I think my strategy'd be rather to have reducers more individually set and unset bools (and reset non-bools) in response to corresponding actions, which I think makes their lifecycle a bit more obvious.